File: apps/backend/src/services/github-auth.ts
import prisma from './db'

const GITHUB_API_BASE = 'https://api.github.com'
const GITHUB_OAUTH_BASE = 'https://github.com/login/oauth'

interface GithubAccessTokenResponse {
  access_token: string
  token_type: string
  scope: string
}

interface GithubUserProfile {
  id: number
  login: string
  name: string | null
  avatar_url: string | null
}

function getRequiredEnv(name: string): string {
  const value = process.env[name]
  if (!value) {
    throw new Error(`Missing required environment variable: ${name}`)
  }
  return value
}

async function githubFetch(input: string, init: RequestInit = {}) {
  return fetch(input, init)
}

export function getGithubOAuthLoginUrl(state: string): string {
  const clientId = getRequiredEnv('GITHUB_CLIENT_ID')
  const redirectUri = getRequiredEnv('GITHUB_REDIRECT_URI')

  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: redirectUri,
    scope: 'read:user read:org',
    state,
  })

  return `${GITHUB_OAUTH_BASE}/authorize?${params.toString()}`
}

export async function exchangeCodeForAccessToken(code: string): Promise<string> {
  const clientId = getRequiredEnv('GITHUB_CLIENT_ID')
  const clientSecret = getRequiredEnv('GITHUB_CLIENT_SECRET')
  const redirectUri = getRequiredEnv('GITHUB_REDIRECT_URI')

  const response = await githubFetch(`${GITHUB_OAUTH_BASE}/access_token`, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      client_id: clientId,
      client_secret: clientSecret,
      code,
      redirect_uri: redirectUri,
    }),
  })

  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to exchange GitHub code: ${response.status} ${text}`)
  }

  const data = (await response.json()) as GithubAccessTokenResponse & { error?: string }
  if (!data.access_token || data.error) {
    throw new Error(`GitHub access token error: ${data.error || 'unknown error'}`)
  }

  return data.access_token
}

export async function fetchGithubUserProfile(accessToken: string): Promise<GithubUserProfile> {
  const response = await githubFetch(`${GITHUB_API_BASE}/user`, {
    headers: {
      Accept: 'application/vnd.github+json',
      Authorization: `Bearer ${accessToken}`,
      'X-GitHub-Api-Version': '2022-11-28',
    },
  })

  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to fetch GitHub user profile: ${response.status} ${text}`)
  }

  return (await response.json()) as GithubUserProfile
}

export async function checkGithubTeamMembership(accessToken: string, login?: string): Promise<boolean> {
  const org = process.env.GITHUB_ORG || 'PCL-Community'
  const teamSlug = process.env.GITHUB_TEAM_SLUG || 'ce-dev'
  const userLogin = login ?? (await fetchGithubUserProfile(accessToken)).login

  const response = await githubFetch(
    `${GITHUB_API_BASE}/orgs/${encodeURIComponent(org)}/teams/${encodeURIComponent(teamSlug)}/memberships/${encodeURIComponent(userLogin)}`,
    {
      headers: {
        Accept: 'application/vnd.github+json',
        Authorization: `Bearer ${accessToken}`,
        'X-GitHub-Api-Version': '2022-11-28',
      },
    }
  )

  if (response.status === 404) {
    return false
  }

  if (!response.ok) {
    const text = await response.text()
    throw new Error(`Failed to check GitHub team membership: ${response.status} ${text}`)
  }

  const membership = (await response.json()) as { state?: string }
  return membership.state === 'active'
}

export async function storeOAuthState(state: string): Promise<void> {
  const expiresAt = new Date(Date.now() + 10 * 60 * 1000)
  await prisma.cache.upsert({
    where: { key: `oauth_state:${state}` },
    update: { value: state, expiresAt },
    create: { key: `oauth_state:${state}`, value: state, expiresAt },
  })
}

export async function validateOAuthState(state: string): Promise<boolean> {
  if (!state) return false

  const record = await prisma.cache.findUnique({
    where: { key: `oauth_state:${state}` },
  })

  if (!record) return false

  if (record.expiresAt && record.expiresAt.getTime() <= Date.now()) {
    await prisma.cache.delete({ where: { key: `oauth_state:${state}` } }).catch(() => undefined)
    return false
  }

  await prisma.cache.delete({ where: { key: `oauth_state:${state}` } }).catch(() => undefined)
  return true
}

File: apps/backend-rs/src/services/github_auth.rs
use reqwest::Client;
use serde::Deserialize;

use crate::config::Config;

#[derive(Debug, Clone, Deserialize)]
pub struct GitHubUser {
    pub id: i64,
    pub login: String,
    pub name: Option<String>,
    pub avatar_url: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GitHubEmail {
    pub email: String,
    pub primary: bool,
    pub verified: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GitHubTeam {
    pub slug: String,
}

pub struct GitHubAuth {
    client: Client,
    cfg: Config,
}

impl GitHubAuth {
    pub fn new(cfg: &Config) -> Self {
        let client = if let Some(proxy_url) = &cfg.github_proxy {
            Client::builder()
                .proxy(reqwest::Proxy::all(proxy_url).unwrap())
                .build()
                .unwrap_or_default()
        } else {
            Client::new()
        };
        Self {
            client,
            cfg: cfg.clone(),
        }
    }

    /// Exchange authorization code for an access token.
    pub async fn exchange_code(&self, code: &str) -> Result<String, String> {
        let params = [
            ("client_id", self.cfg.github_client_id.as_str()),
            ("client_secret", self.cfg.github_client_secret.as_str()),
            ("code", code),
            ("redirect_uri", self.cfg.github_redirect_uri.as_str()),
        ];
        let resp = self
            .client
            .post("https://github.com/login/oauth/access_token")
            .header("Accept", "application/json")
            .form(&params)
            .send()
            .await
            .map_err(|e| format!("request failed: {e}"))?;

        let body: serde_json::Value = resp
            .json()
            .await
            .map_err(|e| format!("parse failed: {e}"))?;

        if let Some(token) = body["access_token"].as_str() {
            Ok(token.to_string())
        } else if let Some(err) = body["error_description"].as_str() {
            Err(err.to_string())
        } else {
            Err("no access_token in response".to_string())
        }
    }

    /// Get authenticated user's profile.
    pub async fn get_user(&self, token: &str) -> Result<GitHubUser, String> {
        let resp = self
            .client
            .get("https://api.github.com/user")
            .header("Authorization", format!("Bearer {token}"))
            .header("User-Agent", "better-apiv2")
            .send()
            .await
            .map_err(|e| format!("request failed: {e}"))?;

        resp.json().await.map_err(|e| format!("parse failed: {e}"))
    }

    /// Check if user is a member of the configured team.
    pub async fn is_team_member(&self, token: &str) -> Result<bool, String> {
        let org = &self.cfg.github_org;
        let slug = &self.cfg.github_team_slug;
        let url = format!(
            "https://api.github.com/orgs/{org}/teams/{slug}/memberships/@@me"
        );

        let resp = self
            .client
            .get(&url)
            .header("Authorization", format!("Bearer {token}"))
            .header("User-Agent", "better-apiv2")
            .send()
            .await
            .map_err(|e| format!("check membership failed: {e}"))?;

        Ok(resp.status().is_success())
    }
}
